test/unit/tool_util/test_tool_linters.py

```diff
@@ -739,7 +764,7 @@
             and "Select parameter of a conditional [select] options have to be defined by 'option' children elements." in x.error_messages
```

Comment:
Just a heads up, I've rebased the branch and I'm rewriting these tests as standard tests. The pytest parameterization makes it very hard to tell what has failed.

Reply:
@mvdbeek thanks. Also just rebased and just wanted to push. Should I wait?

Reply:
Yes, I think that would be better

Reply:
Cool. Just ping me. 

The new linting for assertions will need to get the line and xpath added to the `lint_ctx.error` calls. Also in my rebase I got the repeat linter test duplicated.

Reply:
I think it's better to pass the parameter to the error call, and have the lint_ctx accept a function that gets you the lines. That way you can adapt this to any tool language by providing a different function, and you don't need to call `node_props` everywhere

Reply:
Hmm, so what i've got so far is in https://github.com/galaxyproject/galaxy/compare/dev...mvdbeek:topic/lint-context2?expand=1. Does that look good to you and would you have a chance to finish rewriting the tests ?

Reply:
> Does that look good to you and would you have a chance to finish rewriting the tests ?

Sure. 

> I think it's better to pass the parameter to the error call, and have the lint_ctx accept a function that gets you the lines. 
That way you can adapt this to any tool language by providing a different function, and you don't need to call node_props everywhere

Sounds good. But I have two questions:

- xml tools this seems easy for for source line, but the xpath requires the ETree object .. in external applications like planemo / GLS this will be a different object (built from the same source) .. maybe the xpath lookup won't work then. But I guess xpath is not so important anyway
- in case of cwl tools a parameter would be a dict, or? I have no clue how one can deduce something like the source line from the a dict, i.e. part of the parsed json?

Reply:
Hmm, this was just an idea, maybe it's fine the way it is. I don't know that we really need linting for CWL, the tool loader already does a lot of validation via the schema. If we're investing in that we should first discuss that with the CWL community.